File: src/stats/StatsRecorder.ts
import { repeat } from "../utils/utils";

export const StatsRecorder = _initStatsRecorder();

type StateItem = { date: string, time: number };


function _initStatsRecorder() {
    const empty = '---';
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    let stats: StateItem[] = [];
    const name_state = 'stats_v2';
    load();

    function push(state: StateItem) {
        stats.push(state);
        Storage.set_data(name_state, stats);
    }


    function load() {
        const _stats = Storage.get_data(name_state) as StateItem[];
        stats = _stats ?? create_stats();
        if (stats.length < 10) {
            repeat(4, () => stats.push({ date: empty, time: 0 }));
            Storage.set_data(name_state, stats);
        }
    }



    function create_stats() {
        const stats: StateItem[] = [];
        repeat(10, () => stats.push({ date: empty, time: 0 }));

        Storage.set_data(name_state, stats);
        return stats;

    }

    function get_sorted() {
        return sort(stats);
    }

    function sort(arr: StateItem[]): StateItem[] {
        return arr.sort((a, b) => {
            if (a.time === 0) {
                return 1;
            } else if (b.time === 0) {
                return -1;
            } else {
                return a.time - b.time;
            }

        });
    }

    return {
        push, get_sorted,
    };
}

File: app/stats/StatsRecorder.lua
local ____lualib = require("lualib_bundle")
local __TS__ArraySort = ____lualib.__TS__ArraySort
local ____exports = {}
local _initStatsRecorder
local ____utils = require("utils.utils")
local ____repeat = ____utils["repeat"]
function _initStatsRecorder()
    local load, create_stats, sort, empty, stats, name_state
    function load()
        local _stats = Storage.get_data(name_state)
        stats = _stats or create_stats()
        if #stats < 10 then
            ____repeat(
                4,
                function()
                    local ____temp_0 = #stats + 1
                    stats[____temp_0] = {date = empty, time = 0}
                    return ____temp_0
                end
            )
            Storage.set_data(name_state, stats)
        end
    end
    function create_stats()
        local stats = {}
        ____repeat(
            10,
            function()
                local ____temp_1 = #stats + 1
                stats[____temp_1] = {date = empty, time = 0}
                return ____temp_1
            end
        )
        Storage.set_data(name_state, stats)
        return stats
    end
    function sort(arr)
        return __TS__ArraySort(
            arr,
            function(____, a, b)
                if a.time == 0 then
                    return 1
                elseif b.time == 0 then
                    return -1
                else
                    return a.time - b.time
                end
            end
        )
    end
    empty = "---"
    stats = {}
    name_state = "stats_v2"
    load()
    local function push(state)
        stats[#stats + 1] = state
        Storage.set_data(name_state, stats)
    end
    local function get_sorted()
        return sort(stats)
    end
    return {push = push, get_sorted = get_sorted}
end
____exports.StatsRecorder = _initStatsRecorder()
return ____exports
